components/style/properties/longhand/box.mako.rs

```diff
@@ -1144,39 +1145,453 @@ ${helpers.predefined_type("scroll-snap-coordinate",
     #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
-    enum TranslateKind {
-        Translate,
-        TranslateX,
-        TranslateY,
-        TranslateZ,
-        Translate3D,
+    pub enum SpecifiedRotate {
+        Rotate(SpecifiedAngle),
+        RotateX(SpecifiedAngle),
+        RotateY(SpecifiedAngle),
+        RotateZ(SpecifiedAngle),
+        Rotate3D(CSSFloat, CSSFloat, CSSFloat, SpecifiedAngle),
+    }
+
+    impl fmt::Display for SpecifiedRotate {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            match *self {
+                SpecifiedRotate::Rotate(ref theta) => {
+                    try!(f.write_str("rotate("));
+                        try!(theta.to_css(f));
+                    f.write_str(")")
+                }
+                SpecifiedRotate::RotateX(ref theta)  => {
+                    try!(f.write_str("rotateX("));
+                        try!(theta.to_css(f));
+                    f.write_str(")")
+                }
+                SpecifiedRotate::RotateY(ref theta) => {
+                    try!(f.write_str("rotateY("));
+                        try!(theta.to_css(f));
+                    f.write_str(")")
+                }
+                SpecifiedRotate::RotateZ(ref theta)  => {
+                    try!(f.write_str("rotateZ("));
+                        try!(theta.to_css(f));
+                    f.write_str(")")
+                }
+                SpecifiedRotate::Rotate3D(ax, ay, az, ref theta) => {
+                    try!(f.write_str("rotate3D("));
+                        try!(f.write_str(ax.to_string().as_str()));
+                        try!(f.write_str(", "));
+                        try!(f.write_str(ay.to_string().as_str()));
+                        try!(f.write_str(", "));
+                        try!(f.write_str(az.to_string().as_str()));
+                        try!(f.write_str(", "));
+                        try!(theta.to_css(f));
+                        try!(f.write_str(az.to_string().as_str()));
+                    f.write_str(")")
+                }
+            }
+        }
+    }
+
+    impl ToCss for SpecifiedRotate {
+        fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+            write!(dest, "{}", self)
+        }
     }
 
     #[derive(Clone, Debug, PartialEq)]
     #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
-    enum SpecifiedOperation {
+    pub enum SpecifiedOperation {
         Matrix(SpecifiedMatrix),
-        Skew(specified::Angle, specified::Angle),
-        Translate(TranslateKind,
-                  specified::LengthOrPercentage,
-                  specified::LengthOrPercentage,
-                  specified::Length),
-        Scale(CSSFloat, CSSFloat, CSSFloat),
-        Rotate(CSSFloat, CSSFloat, CSSFloat, specified::Angle),
+        Skew(SpecifiedSkew),
+        Translate(SpecifiedTranslate),
+        Scale(SpecifiedScale),
+        Rotate(SpecifiedRotate),
```

Comment:
I wonder whether it makes much sense to add an additional level here. It seems to me that simply making `SpecifiedOperation` include all the operations could simplify the code, as there doesn't seem to be much code reuse inside the new types.